### DISCUSSION TITLE
git git igit igit

### DISCUSSION THREAD

**Discussion by baiti** [cgeo/cgeo#10320] [General]
at first: pls excuse the title as it  most likely only makes sense in German .. but I just had to ;)

Now down to the wire, here is the situation. I have followed the recipe in the wiki "Working on c:geo for git beginners". A lot of what's there makes sense to me by now but not all, for instance, `git stash --apply` is not a valid command, I believe it should be called `git stash apply`. Not a big deal, but it is certainly confusing to a beginner. Also, I struggled for a long time with `git checkout` vs `git switch`. The wiki doesn't give much rationales, it just says do this then do that and this is probably ok, I don't criticise it. What is of prime importance to me is that I develop a thorough understanding of what's happening underneath when I punch in one of these commands and I must make sure I understand in what state the whole thing is **after** I have executed a git command or a set of such commands via some sort of shell script. (which is what everybody on the team seems to be using as mother's little helpers and which is what I at some point will use too, I am just not there yet).

Let us go into a real life example: I have forked the repo, have created a local copy of the fork, created a branch, coded like hell, done multiple "git commit -am "...." ; git push", created a PR, did many more "git commit -am "..." ; git push". Even though there is now a long list of commits, its all good (e.g. consistent) at this time. My local copy of the fork is in sync with the fork on the server, the PR seems to be consistent as well.

Of course the master repo has progressed a lot over the 4 weeks that have passed now and I guess it would be wise to somehow get in sync with what happened to master in the meantime. I would assume that would greatly simplify the merging of the PR into master lateron when the time has come to do so.

Following the wiki (refreshing your local environment), I would assume the following needs to be done on my local development system and please jump in and warn me if this is wrong: (I haven't executed any of the commands below yet)

`git switch master`
`git pull --rebase upstream master`
`git push origin master`

The sequence is titled: "rebase your local and fork master branch to have the same contents as the upstream master branch".
So, what entities are involved here?

There seem to be:

a local master branch
a fork master branch
an upstream master branch

Why are all these called "master branch" and not just "master"? Is this just splitting hairs? Or is it because everything is ultimatly just "a branch", even master? (Also, one of them should probably be called "origin master". Which one is the origin master?)

And much more important, what's happening by each of the commands above? The first one seems trivial, it just sets some addressing point, sets a focus on something, (I assume it sets the focus on my local copy of the fork to "master").

Now comes the second, `git pull --rebase upstream master`. I assume this goes out to the master repo on github and fetches everything from there as it is now. The command would make much more sense to me if it read `git pull upstream master --rebase` eventhough I know it is most likely semantically equivalent. What it says is pull something from somewhere `git pull upstream master`. I assume this would be a valid command, however with fatal consequences as there needs to be more done that just pulling things. The `--rebase` seems to be the magic here. It seems to change the underlying base onto which my branch (issue_xxxx) was based. I would further assume that this will under the hood trigger some sort of a merge activity, which will in many cases succeed, however, after 4 weeks apart from each other I would further assume that there will be "merge conflicts". Is it all true what I say up to this point?

Now it comes to manual conflict resolution. From what I read it would be best to use the Android Studio for this as it gives me the three window merge thing where the result is in the middle and I can take stuff from either the left side or the right side.

Assume I have fiddled my way through this. In what state is the system afterwards? I would assume the local copy of my fork is now in sync with what is called upstream master plus it has my issue_xxxx branch on top, right? Now my fork on the server side doesn't know of all this yet, which is perhaps what the last command rectifies: `git push origin master`. After this, everything seems to be in sync. ... If I now did a `git switch issue_xxxx ; git status`, everything should be quiet which means good.

But the recipe has more steps to come next:

`git switch <branchname>`
`git rebase master` (<--- which master is this? "upstream master" or "origin master" or some other master?)

Is this the place where the "big merge" takes place and not up there where I thought it would happen? Or does it happen in both passes and BTW: this time "rebase" is a verb and not a command line option ("--rebase"). What did these guys smoke ... really.

Assume I made it down that far. Is it **then** that the `git status` gives me green light and everything is in sync?

What about the PR? I would assume nothing has changed yet to the PR but changes will be necessary. Do I have to do another `git commit -am "just rebased" ; git push`? Is everything "ok" after that?

How about squashing commits, when should that be done? Before everything else mentioned above or at this point where we are now? (And: How exactly is that done?)

I very much appreciate your help and whatever you post here as response I will use as a reference for everything I do from now on. I would use it as a reference to create some of mother's little helpers for my own use too.

Up to this point it is what I'd call the basics. I have one more question though, which is related. Now my PR sits in the queue and probably will be merged at some point in time (I understand why this point may be a loooong way ahead). What if I want to make further changes, strictly for testing new stuff, things that I do not want to throw into the PR. Or maybe into a completely different PR that doesn't exist yet.

Would I create a new branch relative to the one that makes the PR? I would assume I create a new branch on top of the old branch, e.g. `git switch issue_xxxx; git switch -c new_branch`.

Then I do some commits, but the first push should now be `git push -u origin new_branch` otherwise it would screw up the PR? If it is how I think it is, I could now go ahead and do commits and pushes as many as I want and they wouldn't impact the PR? Or do I have to specify the "-u origin <branch_name?" all the time since there is now more than one branch?

It would be a big help if I had clarity on this too. Of course there is another question then further down when it comes to taking some (or all) of the changes in new_branch and merging them into issue_xxxx such that they can go into the PR .... but don't bother that's way down the road, no need to answer right away.

**Comment by SammysHP** [comment]
That are a lot of questions. :) Are you interested in a quick Jitsi meeting (in German)?

**Reply by baiti** [reply]
Yes

**Comment by Narrat** [comment]
Although it will be covered in your meeting, I chime in and write some answers.
I try to be very concise, so I will propably not cover everything:

First point
`git pull` is in general the execution of `git fetch` & `git merge`. There are different merge strategies, how changes are incorporated if someone pulls from a repo. So if there is `git pull --rebase origin master` it can translated to `git fetch origin` & `git rebase origin/master` which rebases your currently checked out branch with the master branch from origin. (I personally avoid the usage of pull, as it does to much for my taste. Sh*t can hit the fan really fast.)

Second:
`git checkout` is a command which is overloaded with functionality (upstream said this). It can be used to switch branches, restore files, checkout a tag or another state. Therefore it got splitted last year. Into a dedicated `git switch` to switch branches and a `git restore` to restore files. And checkout to do different stuff :D (and can still do everything :D)

Third:
Yes, in general master (or main, what is now the new default branch name) is only a branch. Kinda special, as it seen in the default settings as the main branch. But it could be changed if someone wants to.
Every repo can be seen as its own instance. Your local copy, the one you let github host and the repo you cloned. They can diverge and do stuff and can live happily (until someone wants to contribute back).
So if you are working on your PC, your local repo has a branch `master`. If you add another repo from which you fetch changes (for example upstream) then this is `upstream/master`. And so on. `git branch -a` can be used to see every branch your repo knows about (doesn't mean they're switchable)
To see what repository is behind a name you can use `git remote show origin`

Fourth:
`git push origin master` transfers your local master branch to the remote master branch of the repository `origin`.
Without any additional infos it will always push the named branch into remote under the same name.
`git push origin blubb:bla` will push your local `blubb` branch and push it to origin into a branch named `bla` (but forget that for the time being ;))

Fifth:
Rebase takes the branch you name and uses this as base and tries to add your changes from your currently checked out branch.
So if you're on branch `feature1` and issue `git rebase origin/master` it will check, where your branch `feature1` and origin/master diverge, sets `origin/master` as base and then tries to apply the commits, which doesn't exist on `origin/master`.
`git rebase master` uses your local master branch to rebase your current branch onto it.

Sixth:
Squashing commits is `git rebase -i` where -i is for interactive. Don't directly try this is my suggestion. Instead let the before mentioned stuff sink in.

Your idea about the testing stuff is correct. Although I would use the commands `git branch -c pr-branch issueXXXX` & `git switch issueXXXX`.

**Reply by baiti** [reply]
@Narrat ... thank you very much for taking the time to explain and document this. Indeed we covered most of this in the jitsi but it is good to see it written up here. While in the jitsi, we have actually performed those steps interactively and it worked pretty much along the lines documented above. I will now go ahead and extract the whole flow from my bash history and create a well commented script file out of it such that I can review it in case I don't use these commands on a daily basis.

**Comment by baiti** [comment]
After the jitsi I wanted to take the time and cerfully review my bash_history but it turned out I had multiple ssh session open do my development system in the basement. My goal was to construct a mother's little helper that I would use on a regular basis, perhaps daily, do syncrconize my local repo with what's going on in upstream. Now I have to construct this from what I believe I remember from the jitsi and also taking into account @Narrat 's very helpful write up. So here is the goal: Synchronize my local repo with upstream:

`git stash` (otherwise I can't freely do `git switch ......`)
`git switch master && git fetch upstream master && git push` (that should get all three repos in sync wrt master)
`git switch my_feature && git fetch upstream master && git rebase upstream/master`
`git push --force-with-lease` (pushes local repo's current branch to origin's branch, which is my clone)
`git stash apply`

I would expect this runs through no questions asked, however, it can happen that the rebase runs into conflicts, which I then must manually resolve using `git mergetool`.

After all that I can go ahead and do commits and pushes all day long.

Once I have tested this for a while and established confidence I might want to switch to the `--amend` thing to avoid clobbering the commit history.

**Comment by Narrat** [comment]
The write for refreshing your repos with upstream looks good.
Although I will add a few comments :D But those are only intended as small remarks where you can choose if you want to use them.

 1. `git stash`: allows the `-m/--message` option which can be used to add a descriptive title for the stash. Useful if stashes are stored for longer time
 2. `git fetch upstream master`: it is not necessary to specify a branch. (And as sidenote, your `git push` will only work if there is a default repo and branch set. Only mentioned for the case if you stumble upon a non working plain `git push`)
 3. `git switch my_feature && git fetch upstream master && git rebase upstream/master`: if this follows directly after updating your master branch, the `git fetch` isn't really needed. Depends only on how likely it is, that upstream changed in the meantime.
 4. `git stash apply`: if your stash from the beginning is only a shortterm thing, it may be better to use `git stash pop`. Short explanation. You can have more than one stash (see all stashes with `git stash list`). `apply` will apply the specified stash, but will keep it in the stash list. If you do those regulary.. well expect a long list with outdated changes. `pop` on the other hand will apply the stash and remove it from the stash list.

Regarding the commit history. `git commit --amend` can come in handy, but usage on a feature branch? Following example. Maybe you plan a rather large change and you do it in various logical steps. If you amend every change to your first commit you will have more work if you discover, that one of the changes broke something or should be completly redone from scratch. If every logical change is a commit and you discover something, then you can work with `git reset` or `git revert` to undo those specific changes. If you think your feature is ready, then the interactive rebase (`git rebase -i`) will come in handy. There you can decide which commits shall stay, which commits should be added to others or entirely be dropped.

**Reply by baiti** [reply]
> The write for refreshing your repos with upstream looks good.
> Although I will add a few comments :D But those are only intended as small remarks where you can choose if you want to use them.

**very** much appreciated. I know it will be a long way until I may be ready for the very first "git meister belt". If I ever make it thus far.

> 
> 1. `git stash`: allows the `-m/--message` option which can be used to add a descriptive title for the stash. Useful if stashes are stored for longer time
> 2. `git fetch upstream master`: it is not necessary to specify a branch. (And as sidenote, your `git push` will only work if there is a default repo and branch set. Only mentioned for the case if you stumble upon a non working plain `git push`)

Since I currently have only one branch and used `git push -u origin <branchname>` for the first push I figured this would work, knowing that I have to be more explicit as soon as I am working on more than one branch at a time.


> 3. `git switch my_feature && git fetch upstream master && git rebase upstream/master`: if this follows directly after updating your master branch, the `git fetch` isn't really needed. Depends only on how likely it is, that upstream changed in the meantime.

Good point. I thought so too but wanted to be on the safe side. Upstream in the given case only changes if someone on the team merges a PR which is not really a high frequency activity. At least it wasn't in the time period  I am monitoring progress.

> 4. `git stash apply`: if your stash from the beginning is only a shortterm thing, it may be better to use `git stash pop`. Short explanation. You can have more than one stash (see all stashes with `git stash list`). `apply` will apply the specified stash, but will keep it in the stash list. If you do those regulary.. well expect a long list with outdated changes. `pop` on the other hand will apply the stash and remove it from the stash list.

`git stash pop` was mentioned in the jitsi on the side. Just did a `git stash list` and it was indeed rather long and always the same stash. So I did `git stash drop` until the list was empty.

My strong believe in symmetry tells me I want a `git stash push` and a `git stash pop`. As it seems, just doing a `git stash` is really a short hand for `git stash push` but not specifying the push made me not even **see** that there is a pop.

> 
> Regarding the commit history. `git commit --amend` can come in handy, but usage on a feature branch? Following example. Maybe you plan a rather large change and you do it in various logical steps. If you amend every change to your first commit you will have more work if you discover, that one of the changes broke something or should be completly redone from scratch. If every logical change is a commit and you discover something, then you can work with `git reset` or `git revert` to undo those specific changes. If you think your feature is ready, then the interactive rebase (`git rebase -i`) will come in handy. There you can decide which commits shall stay, which commits should be added to others or entirely be dropped.

So far my workflow followed the "release early, release often" paradigm which I translated into a frequent `git commit -am ".." ; gut push` sequence and since I also created the PR early in the game, the PR was updated all the time and you can see a growing list of commits in the PR. What you say makes a lot of sense to me so I think I will stay away from the --amend thing. My current understanding is that it wouldn't do any good to my workflow anyway at least not as long as I always do the `git commit ; git push`. I believe the git commit --amend only helps if I have done commits but not (yet) pushed those? In other words, once pushed it is too late for the --amend, right?

**Reply by Narrat** [reply]
> I know it will be a long way until I may be ready for the very first "git meister belt". If I ever make it thus far.

Git can indeed be intimidating in the beginning. But that will change ;)

> Since I currently have only one branch and used git `push -u origin <branchname>` for the first push I figured this would work, knowing that I have to be more explicit as soon as I am working on more than one branch at a time.

It will work. And if you know about possible upcoming quirks, when your work with git expands, then everything is good :) I'm just from the faction, which does stuff the long way in the beginning. Before setting options for convenience :D

> As it seems, just doing a `git stash` is really a short hand for `git stash push` (...)

Mostly, yes. It does behave a little bit differently if the `push` is omitted. If I remember correctly some options are not available. The man-page can probably tell what is different.

> So far my workflow followed the "release early, release often" paradigm which I translated into a frequent `git commit -am ".." ; gut push` sequence and since I also created the PR early in the game, the PR was updated all the time and you can see a growing list of commits in the PR. What you say makes a lot of sense to me so I think I will stay away from the --amend thing. My current understanding is that it wouldn't do any good to my workflow anyway at least not as long as I always do the git commit ; git push. I believe the `git commit --amend` only helps if I have done commits but not (yet) pushed those? In other words, once pushed it is too late for the --amend, right?

A very good paradigm :D One remark: If you haven't done it, think about using `git add` & `git commit` instead of `git commit -a`. It allows for more control which changes should be staged for a commit.
In general yes, this is the intended workflow with `--amend`. If you committed something but didn't push somewhere else. And shortly (or even directly) after you made the commit you notice that something wasn't included. Add, amend and finally push the changes.
But the answer can also be: It depends. On the fact if you are the only user, where an amended commit lands and how frequently it is used.
Because even if you pushed the changes before amending something, you can always use a force-push to overwrite what is at a remote location where you have access to. But this treads into dangerous territory :D If it is just force-pushing a branch where you work alone and don't have different workplaces.. Not that much of a problem, although it breaks the review process of github if I remember it correctly. If some other person works with you on the branch.. Well, this could then invoke some mean comments :D And even worse if `master` is force-pushed with a break in history due to an amended commit or rewritten history. Everyone with a clone of the repo which got updated to the wrong time need to fix their local repo so it has again a consistent history.

**tl;dr:** Yes, `--amend` works best locally, before pushing changes somewhere else.

**Reply by Lineflyer** [reply]
The usual workflow for PRs to c:geo is (at least for me):
1. Create a new branch based on the branch you would like to have as base (typically `master`) and give it a name like ìssues_xyz`
2. Add your work and commit it
3. Push the branch to remote and create a PR
4. If it happens now, that you want to change/amend the PR, then make another commit on that branch and push it to update the PR
5. Ideally I want that my PR to stay with a single commit (in order not to clutter the history), so instead of 4. I do a `git rebase -i HEAD~2` and squash the new commit into the existing one. Afterwards I use a `git push --force` to forcefully update the branch and my PR.

Sorry if this was all already mentioned above, I did not read all comments fully.

**Comment by baiti** [comment]
The more I think about it, the more I believe we should have a sandbox environment which matches the cgeo workflow as close as possible. I will see whether I can setup something like this myself or with some helps from the resident experts.

**Reply by Lineflyer** [reply]
You can just fork c:geo and play around as you like on that fork. Thats also how I started into the world of Git.
Or plainly create a new repository on Github and use it as a playground for some push/pull/merge practice.

**Reply by baiti** [reply]
> You can just fork c:geo and play around as you like on that fork. Thats also how I started into the world of Git.

I had thought that too but didn't pursue it because I concluded I'd need to register an additional userid with github as I cannot create a second fork named "baiti/cgeo" and I didn't want to take the risk to screw up my existing fork at least not while I have a PR "in progress" and I cannot afford to loose that work as it took me days to write it.

> Or plainly create a new repository on Github and use it as a playground for some push/pull/merge practice.

That is an approach that seems plausible, except that I would also need at least two userids at github and each would do the fork and clone and commit and push and create PR and sync the local environment and then one of it would have write access to upstream / master and ultimately merge the PR. I was too far away from grasping all of this flow but I believe I am now at a point where I could actually pursue that route.

**Reply by SammysHP** [reply]
You can just use another local repository as a remote. Instead of cloning a git/ssh url clone a file path.

It is also easy to create a second fork. Create a new project on GitHub and push into it.

One benefit of git is that it's so easy to have many remotes without needing special server software.

**Comment by baiti** [comment]
I have a question coming up soon, now that my PR is almost ready for merge. Once it is merged, I do:

```
git switch master
git fetch upstream
```

Ideally I should be able to delete my branch, `minimum_ALC` in this case, except that I have created another branch based on `minimum_ALC`, called `full_ALC`. What happens to `full_ALC` if I delete `minium_ALC`? I would assume this would not be a good idea.

Would it be a better idea to do:

```
git switch master
git fetch upstream
git switch full_ALC
git rebase -i upstream master
```
can I afterwards do `git branch -d minimum_ALC`?

**Comment by Lineflyer** [comment]
Once your PR is merged you usually don't need the corresponding branch any more.
Deleting a branch does not influence any other branch in your repo. You can keep using them normally and pushing or rebasing them as needed.

BTW:
I also preserved your original full_ALC PR in a branch on the cgeo repository.
You could also fetch/pull this from upstream use it and (if desired) open a PR to further improve ir or rebase it to keep it up to date (if interested...right now we don't know if this might get merged later on).

If its interesting for anyone I could also establish a job on the CI to build a signed nightliy from that....mmh

**Reply by baiti** [reply]
> Once your PR is merged you usually don't need the corresponding branch any more.
> Deleting a branch does not influence any other branch in your repo. You can keep using them normally and pushing or rebasing them as needed.

Not even if the branch to be deleted is the parent of a child? ... Which I clearly want to preserve.

> 
> BTW:
> I also preserved your original full_ALC PR in a branch on the cgeo repository.
> You could also fetch/pull this from upstream use it and (if desired) open a PR to further improve ir or rebase it to keep it up to date (if interested...right now we don't know if this might get merged later on).

full_ALC does not only mean more capabilities on ALC such as stages, it also means that it has the `Ilogin, ICrentantials` capabilities and is ready for doing authorization. Should we be able to convince GC HQ to give us the capability to render the "Found State" of ALCs, something they offer in the Geocaching App, then we may need to revitalize (and improve) some of the functionality in full_ALC. See my post on the team channel for more details.

> 
> If its interesting for anyone I could also establish a job on the CI to build a signed nightliy from that....mmh

I wouldn't suggest to do this at this time. The other one, still has the relevant capabilities ..... But it may be an option further down the road. Thanks for offering it.

**Reply by baiti** [reply]
> If its interesting for anyone I could also establish a job on the CI to build a signed nightliy from that....mmh

Strictly from the perspective of understanding the process: Wouldn't that mean that the branch in question would need to be rebased on master on a regular basis? That couldn't possibly be a task in CI right? There would be merge conflicts not resolvable by Mr. Jenkins I'd assume.

**Reply by fm-sys** [reply]
> Not even if the branch to be deleted is the parent of a child?
 
A branch simply is a pointer to a specific commit. But if you delete a branch only this pointer is deleted, not the commit(s) itself.

**Reply by baiti** [reply]
> > Not even if the branch to be deleted is the parent of a child?
> 
> A branch simply is a pointer to a specific commit. But if you delete a branch only this pointer is deleted, not the commit(s) itself.

@fm-sys  ... That is a good abstraction. Thanks much.
